src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK013ThreadAffinitizedMEFConstruction.cs

```diff
@@ -0,0 +1,328 @@
+                    {
+                        var semanticModel = context.Compilation.GetSemanticModel(statement.SyntaxTree);
+                        var expressionStatement = statement as ExpressionStatementSyntax;
+                        if (expressionStatement?.Expression is InvocationExpressionSyntax invocationExpression)
+                        {
+                            var symbolInfo = semanticModel.GetSymbolInfo(invocationExpression.Expression, context.CancellationToken).Symbol;
+                            this.AnalyzeMemberWithinContext(symbolInfo.ContainingType, symbolInfo, context, statement.GetLocation());
+                        }
+                    }
+                }
+            }
+        }
+
+        internal void AnalyzeType(SymbolAnalysisContext context, INamedTypeSymbol exportAttributeType)
+        {
+            // Ensure the symbol is a named type
+            if (context.Symbol is not INamedTypeSymbol namedTypeSymbol)
+            {
+                return;
+            }
+
+            // Check if the class is decorated with the Export attribute
+            if (!namedTypeSymbol.GetAttributes().Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, exportAttributeType)))
+            {
+                return;
+            }
+
+            // Find constructors with no parameters
+            var parameterlessConstructors = namedTypeSymbol.Constructors
+                .Where(ctor => ctor.Parameters.Length == 0);
+
+            foreach (var parameterlessConstructor in parameterlessConstructors)
+            {
+                this.AnalyzeMethodContents(context, parameterlessConstructor);
+            }
+
+            // Find fields and properties with initializers
+            var fieldsWithInitializers = namedTypeSymbol.GetMembers()
+                .OfType<IFieldSymbol>()
+                .Where(field => field.DeclaringSyntaxReferences
+                    .Select(reference => reference.GetSyntax())
+                    .OfType<VariableDeclaratorSyntax>()
+                    .Any(declarator => declarator.Initializer != null));
+
+            foreach (var field in fieldsWithInitializers)
+            {
+                var declaratorSyntax = field.DeclaringSyntaxReferences
+                    .Select(reference => reference.GetSyntax())
+                    .OfType<VariableDeclaratorSyntax>()
+                    .FirstOrDefault(declarator => declarator.Initializer != null);
+
+                if (declaratorSyntax?.Initializer?.Value is ExpressionSyntax initializerExpression)
+                {
+                    var semanticModel = context.Compilation.GetSemanticModel(initializerExpression.SyntaxTree);
+
+                    // Traverse the initializer expression to analyze all symbols
+                    foreach (var descendantNode in initializerExpression.DescendantNodesAndSelf())
```

Comment:
What's the best way to limit this analyzer to finding just one diagnostic?

I suppose that if I filter the traversed nodes here, I can get to single diagnostic, at a risk of having false positives in filtered cases. I also considered using a hashset of traversed symbols instead.

The test case contains
```object o = Microsoft.VisualStudio.Shell.UIContext.FromUIContextGuid(System.Guid.Empty);```
and in current form, the analyzer reports diagnostics at
1. `Microsoft.VisualStudio.Shell.UIContext.FromUIContextGuid`
2. `Microsoft.VisualStudio.Shell.UIContext.FromUIContextGuid(System.Guid.Empty)`
3. `FromUIContextGuid`

Reply:
Have you considered an IOperation based search instead of a syntax node based search? That will perform better and I suspect be simpler.

Reply:
I've onboarded onto the IOperation paradigm